### ISSUE TITLE
Using Raspberry Pi Camera with Wifi and Picam

### ISSUE THREAD

**Issue by nelsonic** [dwyl/learn-nerves#5]
@pdgonzalez872 wrote this post in 2018. it looks like a good starting point:
https://medium.com/@pdgonzalez872/using-nerves-with-rpi3-wifi-and-picam-3b04cee78d22
Code + full walkthrough: https://github.com/pdgonzalez872/nerves_rpi3_wifi_picam

It would be _great_ to:
+ [ ] See if everything still works in 2020
+ [ ] Update it to use the latest hardware (_Raspberry Pi 4 **and/or** Zero W_)
+ [ ] Test it with streaming images/video to a screen either on a separate device or connected to the Pi.
+ [ ] Write it up!

The practical applications are:
a. Home security camera (external): https://github.com/dwyl/home/issues/130
b. Monitoring Plants/Fish/Grow System for Aquaponics https://github.com/dwyl/learn-aquaponics/issues
c. OctoPi to monitor/record 3D printer activity - https://octoprint.org | https://youtu.be/ZxHf1LqPiPI
d. GrannyCam - create a secure stream using [RTSP](https://en.wikipedia.org/wiki/Real_Time_Streaming_Protocol) / [WebRTC](https://en.wikipedia.org/wiki/WebRTC)
_obviously_ we wouldn't advise _anyone_ to have a camera streaming _publicly_ on the Internet! But with [**`auth`**](https://github.com/dwyl/auth) it could be a fun little project to create a bi-directional stream between grandparents and grandchildren who are otherwise separated by thousands of miles and only get updates when exhausted new parents remember to send them ... We know a couple of Grandmothers who would be addicted to this 24/7. 😜   ... Baby photos/videos are basically Crack for Grandmothers, they just can't get enough. 🙄 

I have the V2.1 Camera and a RPI3 (_the exact components used in this tutorial_), so I can try and get it working.
But we could consider getting a more advanced camera (with a variable focus lens): https://www.amazon.co.uk/dp/B07R4JH2ZV
![image](https://user-images.githubusercontent.com/194400/86794435-7034ea00-c064-11ea-9d7d-9f9b30846242.png)

@th0mas are you interested in helping out with this after you have finished your current work on Access Control?

**Comment by pdgonzalez872** [comment]
Please let me know and I'll edit the article to show your findings! This is excellent!

**Comment by nelsonic** [comment]
@th0mas as noted in https://github.com/dwyl/smart-home-security-system/issues/10 we would like to use a Camera Connected to the Pi to detect motion and switch on the Welcome Display. Which camera do you think would be appropriate for that? (Please do a some Amazon research and share a link so we can send you one today). Thanks!

**Comment by t0mhaines** [comment]
For your use case, I'd use https://www.amazon.co.uk/LABISTS-Official-Camera-Surrordings-Raspberry/dp/B07VT2KGQC:

<img width="1439" alt="Screenshot 2020-07-10 at 10 20 20" src="https://user-images.githubusercontent.com/9089056/87138959-23852500-c297-11ea-8df0-c51d98c51234.png">

Which Is an Infrared  version of the Standard Pi camera.

There's a few problems that we'd need to overcome to tick all the boxes however:

- Motion detection is a *hard* problem. Theres the [Motion](https://github.com/Motion-Project/motion) library that will help with this, but we'll need to create a custom Nerves image and write an Elixir wrapper round it to use.

- Streaming Video is hard. The Elixir Picam library only supports MJPEG video streaming, which, while fine for local network use, is too inefficient to use over WAN. 

  Streaming using the build in libraries then piping to FFMPEG seems to be the way to go? https://projects.raspberrypi.org/en/projects/infrared-bird-box/10

[Membrane](https://www.membraneframework.org/) Looks promising but doesn't have Native Picam support *yet*

**Comment by nelsonic** [comment]
Yeah, totally understand that doing motion detection is definitely not _easy_. (_hence making it a stretch goal_). 👍 
The only reason I thought it was doable is because a couple of people have managed to do it using Elixir/Nerves.

Arto Bendiken (@artob) gave a _superb_ presentation on Home Security with Elixir and Nerves 
where he demonstrates using a PI + Cam to do **facial detection**:
[![image](https://user-images.githubusercontent.com/194400/87145741-af03b380-c2a1-11ea-8133-cc48477e89eb.png)](https://youtu.be/oyNSmhkS7Dw)
https://youtu.be/oyNSmhkS7Dw - the slides are: https://speakerdeck.com/arto/building-a-home-security-system-with-elixir-and-nerves and corresponding blog post is: https://ar.to/2019/02/talk-at-pivorak-conf-2
But sadly, the **`code`** does not _appear_ to be Open Source ... 
maybe I'm not looking in the right place: https://github.com/artob?tab=repositories ... 🤷 
We could just _ask_ Arto really nicely if he can share some of his work to save time. 

Similarly Jacqueline Manzi's (@JacquiManzi) _excellent_ ElixirConf 2019 talk "Breaking Into Nerves" does **motion detection**:
[![image](https://user-images.githubusercontent.com/194400/87146545-05252680-c2a3-11ea-8407-a96e1d32d9cb.png)](https://youtu.be/IU8epH8FAI4)
https://youtu.be/IU8epH8FAI4

The code appears to be: https://github.com/JacquiManzi/jacqui_manzi_elixir_conf_breaking_into_nerves
And the motion worker looks like a good starting point for tackling the challenge of motion detection (without face recognition): https://github.com/JacquiManzi/jacqui_manzi_elixir_conf_breaking_into_nerves/blob/master/lib/hello_nerves/motion/worker.ex
